### ISSUE TITLE
[dev] command to build and test docs snippets

### ISSUE THREAD

**Issue by cohix** [suborbital/subo#22]
Would be great to have some Go files in a directory, and be able to add template markers in documentation markdown such that running `subo docs build` would build and run all of the Go code to ensure it's still correct, then insert the appropriate snippets into the documentation.

There should also be a `subo docs test` command wherein it tests the snippets but doesn't re-generate the docs, to be run in CI.

**Comment by denopink** [comment]
Sounds pretty cool!
In terms of implementation, were you thinking of something like the following
![image](https://user-images.githubusercontent.com/12867159/141393444-66a55cc7-506c-4051-891d-2d17e2ccb2b3.png)

Where the file greetings.go
```golang
/*
Package greetings
*/

// Hello, {{.foo}}.
func FuncPathAndName(f interface{}) string {
	return runtime.FuncForPC(reflect.ValueOf(f).Pointer()).Name()
}
```
Would produce the following
```golang
/*
Package greetings
*/

// Hello, bar.
func FuncPathAndName(f interface{}) string {
	return runtime.FuncForPC(reflect.ValueOf(f).Pointer()).Name()
}
```

**Comment by cohix** [comment]
@denopink It's actually the inverse! I'd like to be able to take a bunch of markdown files (like the ones that go into our various GitBook docs pages like https://atmo.suborbital.dev) and be able to insert Go code snippets into them :)

This is so that our getting started guides and other docs pages maintain up to date and correct code snippets in them :)

**Comment by denopink** [comment]
@cohix oh(!), that makes more sense. When I read `add template markers in documentation markdown`, I thought the documentation markdown (where we're looking to insert the code/snippets) was referring to the actual go documentation in the go file. Thinking back... that wouldn't be very useful!

So it may look like something like this
![image](https://user-images.githubusercontent.com/12867159/141665320-af725858-a2b5-4dc4-b638-ec589e24a3df.png)

If no one else is working on this, then I would love to take it on :)

**Comment by cohix** [comment]
@denopink That looks about right! I don't think we need to create an AST of the code to start, I think a good first step would be to just insert the entire `.go` file into the markdown, and then we can look later at adding custom markers in the Go files to indicate seperate snippets.

In the markdown, I'd love to see something like:
```
{{ snippet somefile.go }}
```
Which then inserts the entirety of `somefile.go` into the markdown, within triple-backticks

Please feel free to take it on!

**Comment by denopink** [comment]
@cohix thx! Regarding the go build/test feature, I have been familiarizing myself with the code base and I saw that we have a `builder` api. In the `builder` api, I saw the `BuildWithToolchain`/`dockerBuildForLang` apis.

Although the command being executed in the docker container is hard coded in `dockerBuildForLang` (builder.go, line 200), (for the build/test feature) I'm assuming we'll either want to
1) leverage the apis used in `BuildWithToolchain`/`dockerBuildForLang` to create a general build/test api for the `subo docs {build | test}` feature or
2) remove the hard coded command being executed in the docker container in `dockerBuildForLang` (builder.go, line 200) with another b.Context:

https://github.com/suborbital/subo/blob/750b9c98a15b54d74effdbf91bd0baf9c800a4af/builder/builder.go#L192-L200

**Comment by denopink** [comment]
@cohix or should we simply do a `exec.Command("go", "build")` for the build/test feature? I originally thought we would want to remove any local dependencies for building the go code, so running 'go build' in a docker container seemed like the natural solution.

**Comment by cohix** [comment]
@denopink just checking in to see if this is still something you'd like to tackle?

**Comment by denopink** [comment]
Yeah, it's a cool application of text templates. I meant for this to be my first issue ticket, but then I got distracted by the other tickets! 🙃

But since it's a good starting point for new comers, I don't mind passing it along if one of the externships (or anyone) was looking for a good first issue 🙂

**Comment by denopink** [comment]
@jagger27 @cohix looks like this issue has a `wait` tag, should I hold off on a pull request for this?

**Comment by cohix** [comment]
@denopink not sure why it did... have at it! 🙂

**Comment by denopink** [comment]
can we close this since we merged #158 ?
